diff --git a/utils/v4l2-ctl/v4l2-ctl-common.cpp b/utils/v4l2-ctl/v4l2-ctl-common.cpp
index 47f5da1..e712679 100644
--- a/utils/v4l2-ctl/v4l2-ctl-common.cpp
+++ b/utils/v4l2-ctl/v4l2-ctl-common.cpp
@@ -785,14 +785,19 @@ static bool parse_subset(char *optarg)
 
 static bool parse_next_subopt(char **subs, char **value)
 {
-	static char *const subopts[] = {
-	    NULL
-	};
-	int opt = getsubopt(subs, subopts, value);
+	char *stmp = *subs;
+	*value = NULL;
+
+	if (*subs) {
+		*subs = strchr(stmp, ',');
+		if (*subs)
+			*(*subs)++ = 0;
+		else
+			*subs = stmp + strlen(stmp);
 
-	if (opt < 0 || *value)
+		*value = stmp;
 		return false;
-	fprintf(stderr, "Missing suboption value\n");
+	}
 	return true;
 }
 
