From 45e3876f009f375218716ed6310240c361feb87f Mon Sep 17 00:00:00 2001
From: luwinkey <luwinkey@allwinnertech.com>
Date: Thu, 21 Jan 2021 13:12:19 +0800
Subject: [PATCH] refused some devices re-connect when connected.

Signed-off-by: luwinkey <luwinkey@allwinnertech.com>
---
 profiles/input/device.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/profiles/input/device.c b/profiles/input/device.c
index d89da2d..963d5d0 100644
--- a/profiles/input/device.c
+++ b/profiles/input/device.c
@@ -340,7 +340,7 @@ static gboolean intr_watch_cb(GIOChannel *chan, GIOCondition cond, gpointer data
 	btd_service_disconnecting_complete(idev->service, 0);
 
 	/* Enter the auto-reconnect mode if needed */
-	input_device_enter_reconnect_mode(idev);
+	// input_device_enter_reconnect_mode(idev);
 
 	return FALSE;
 }
-- 
2.29.0

