From d9743dd7c4bb69143c0c1c4fe16fa165d449cef9 Mon Sep 17 00:00:00 2001
From: caizepeng <caizepeng@allwinnertech.com>
Date: Tue, 11 Jan 2022 11:28:33 +0800
Subject: [PATCH] fix input/hoglib reportid erroneously remove and add

Signed-off-by: caizepeng <caizepeng@allwinnertech.com>
---
 profiles/input/hog-lib.c | 37 +++++++++++++++++++++++++------------
 1 file changed, 25 insertions(+), 12 deletions(-)

diff --git a/profiles/input/hog-lib.c b/profiles/input/hog-lib.c
index 9c5c814..bf0e997 100644
--- a/profiles/input/hog-lib.c
+++ b/profiles/input/hog-lib.c
@@ -311,10 +311,17 @@ static void report_value_cb(const guint8 *pdu, guint16 len, gpointer user_data)
 	buf = ev.u.input.data;
 
 	if (hog->has_report_id) {
-		buf[0] = report->id;
-		len = MIN(len, sizeof(ev.u.input.data) - 1);
-		memcpy(buf + 1, pdu, len);
-		ev.u.input.size = ++len;
+		// add report id prefix only if non-zero (which is reserved)
+		if (report->id != 0) {
+			buf[0] = report->id;
+			len = MIN(len, sizeof(ev.u.input.data) - 1);
+			memcpy(buf + 1, pdu, len);
+			ev.u.input.size = ++len;
+		} else {
+			len = MIN(len, sizeof(ev.u.input.data) - 1);
+			memcpy(buf, pdu, len);
+			ev.u.input.size = len;
+		}
 	} else {
 		len = MIN(len, sizeof(ev.u.input.data));
 		memcpy(buf, pdu, len);
@@ -793,7 +800,9 @@ fail:
 static void get_report_cb(guint8 status, const guint8 *pdu, guint16 len,
 							gpointer user_data)
 {
-	struct bt_hog *hog = user_data;
+	//struct bt_hog *hog = user_data;
+	struct report *report = user_data;
+	struct bt_hog *hog = report->hog;
 	struct uhid_event rsp;
 	int err;
 
@@ -823,13 +832,14 @@ static void get_report_cb(guint8 status, const guint8 *pdu, guint16 len,
 	--len;
 	++pdu;
 	if (hog->has_report_id && len > 0) {
-		--len;
-		++pdu;
+		rsp.u.get_report_reply.size = len + 1;
+		rsp.u.get_report_reply.data[0] = report->id;
+		memcpy(&rsp.u.get_report_reply.data[1], pdu, len);
+	} else {
+		rsp.u.get_report_reply.size = len;
+		memcpy(rsp.u.get_report_reply.data, pdu, len);
 	}
 
-	rsp.u.get_report_reply.size = len;
-	memcpy(rsp.u.get_report_reply.data, pdu, len);
-
 exit:
 	rsp.u.get_report_reply.err = status;
 	err = bt_uhid_send(hog->uhid, &rsp);
@@ -860,7 +870,8 @@ static void get_report(struct uhid_event *ev, void *user_data)
 
 	hog->getrep_att = gatt_read_char(hog->attrib,
 						report->value_handle,
-						get_report_cb, hog);
+						get_report_cb, report);
+
 	if (!hog->getrep_att) {
 		err = ENOMEM;
 		goto fail;
@@ -970,8 +981,10 @@ static void report_map_read_cb(guint8 status, const guint8 *pdu, guint16 plen,
 		if (get_descriptor_item_info(&value[i], vlen - i, &ilen,
 								&long_item)) {
 			/* Report ID is short item with prefix 100001xx */
-			if (!long_item && (value[i] & 0xfc) == 0x84)
+			if (!long_item && (value[i] & 0xfc) == 0x84) {
 				hog->has_report_id = TRUE;
+				DBG("hog->has_report_id = TRUE;value[i] %02X",value[i]);
+			}
 
 			DBG("\t%s", item2string(itemstr, &value[i], ilen));
 
-- 
2.29.0

